### PR TITLE
 Implement inspector support for our desktop platforms.

### PR DESCRIPTION
**Pull request by darktears**
This include Windows and Linux. It enables the right click to inspect an element. It will launch a new window where the inspector will be loaded.

This is a different workflow than Android where it's remote debugging over USB where you debug on your desktop Chrome.

I removed xwalk_devtools_delegate, these files were used in xwalk_content_browser_client but beside returning a instance to ContentClient it was doing nothing. I fail to see what it was doing before. Anyway it's now covered by the new implementation.

BUG=XWALK-4858

### CODE REVIEW

**Comment by darktears** [comment]
@pozdnyakov r?

anyone r?

File: runtime/browser/devtools/xwalk_devtools_frontend.cc [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+void XWalkDevToolsFrontend::Close() {
+  runtime_window_->Close();
+}
+
+void XWalkDevToolsFrontend::DisconnectFromTarget() {
+  if (!agent_host_)
+    return;
+  agent_host_->DetachClient();
+  agent_host_ = NULL;
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: nullptr

File: runtime/browser/devtools/xwalk_devtools_frontend.cc [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    NativeAppWindowDesktop* runtime_window,
+    content::WebContents* inspected_contents)
+    : content::WebContentsObserver(runtime_window->web_contents()),
+      runtime_window_(runtime_window),
+      inspected_contents_(inspected_contents),
+      weak_factory_(this) {
+}
+
+XWalkDevToolsFrontend::~XWalkDevToolsFrontend() {
+  for (const auto& pair : pending_requests_)
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: linked_ptr could be used instead

File: runtime/browser/devtools/xwalk_devtools_frontend.cc [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    agent_host_->DetachClient();
+  delete this;
+}
+
+void XWalkDevToolsFrontend::HandleMessageFromDevToolsFrontend(
+    const std::string& message) {
+  if (!agent_host_)
+    return;
+  std::string method;
+  base::ListValue* params = NULL;
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: nit: nullptr

File: runtime/browser/devtools/xwalk_devtools_frontend.cc [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  base::DictionaryValue* dict = NULL;
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: ditto

File: runtime/browser/devtools/xwalk_devtools_frontend.cc [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+void XWalkDevToolsFrontend::WebContentsDestroyed() {
+  if (agent_host_)
+    agent_host_->DetachClient();
+  delete this;
+}
+
+void XWalkDevToolsFrontend::HandleMessageFromDevToolsFrontend(
+    const std::string& message) {
+  if (!agent_host_)
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: when do we expect this?

File: runtime/browser/devtools/xwalk_devtools_frontend.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  static XWalkDevToolsFrontend* Show(content::WebContents* inspected_contents);
+
+  void Activate();
+  void Focus();
+  void InspectElementAt(int x, int y);
+  void Close();
+
+  void DisconnectFromTarget();
+
+  NativeAppWindowDesktop* frontend_shell() const { return runtime_window_; }
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.h]: either 'const NativeAppWindowDesktop* frontend_shell() const' or 'NativeAppWindowDesktop* frontend_shell()'

File: runtime/browser/devtools/xwalk_devtools_frontend.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                          const base::Value* arg2,
+                          const base::Value* arg3);
+
+ protected:
+  XWalkDevToolsFrontend(NativeAppWindowDesktop* runtime_window,
+                        content::WebContents* inspected_contents);
+  ~XWalkDevToolsFrontend() override;
+
+  void OnRuntimeClosed(Runtime* runtime) override;
+  void OnNewRuntimeAdded(Runtime* runtime) override {};
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.h]: cannot it be omitted?

File: runtime/browser/devtools/xwalk_devtools_frontend.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  // content::DevToolsAgentHostClient implementation.
+  void AgentHostClosed(
+      content::DevToolsAgentHost* agent_host, bool replaced) override;
+  void DispatchProtocolMessage(content::DevToolsAgentHost* agent_host,
+                               const std::string& message) override;
+  base::DictionaryValue* preferences() { return &preferences_; }
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.h]: why not return base::DictionaryValue& ?

File: runtime/browser/devtools/xwalk_devtools_manager_delegate.cc [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+};
+
+scoped_ptr<DevToolsHttpHandler::ServerSocketFactory>
+CreateSocketFactory() {
+  const base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+  // See if the user specified a port on the command line (useful for
+  // automation). If not, use an ephemeral port by specifying 0.
+  uint16 port = 0;
+  if (command_line.HasSwitch(switches::kRemoteDebuggingPort)) {
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: can we leave cmd line parsing in main parts?

File: runtime/browser/devtools/xwalk_devtools_manager_delegate.cc [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        command_line.GetSwitchValueASCII(switches::kRemoteDebuggingPort);
+    if (base::StringToInt(port_str, &temp_port) &&
+        temp_port > 0 && temp_port < 65535) {
+      port = static_cast<uint16>(temp_port);
+    } else {
+      DLOG(WARNING) << "Invalid http debugger port number " << temp_port;
+    }
+  }
+  return scoped_ptr<DevToolsHttpHandler::ServerSocketFactory>(
+      new TCPServerSocketFactory("127.0.0.1", port));
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: worth having a const variable

File: runtime/browser/devtools/xwalk_devtools_manager_delegate.cc [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    : browser_context_(browser_context) {
+}
+
+XWalkDevToolsManagerDelegate::~XWalkDevToolsManagerDelegate() {
+}
+
+base::DictionaryValue* XWalkDevToolsManagerDelegate::HandleCommand(
+    content::DevToolsAgentHost* agent_host,
+    base::DictionaryValue* command) {
+  return NULL;
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: nullptr

File: runtime/browser/devtools/xwalk_devtools_frontend.cc (hunk 5 above)
Comment (darktears) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: override of content::DevToolsFrontendHost::Delegate implementation if it's what you're asking. Pure virtual method, needs to be reimplemented.

File: runtime/browser/runtime.cc [hunk 12]
```diff
@@ -312,6 +312,12 @@ void Runtime::DidDownloadFavicon(int id,
     ui_delegate_->UpdateIcon(app_icon_);
 }
 
+bool Runtime::HandleContextMenu(const content::ContextMenuParams& params) {
+  if (ui_delegate_)
+    return ui_delegate_->GetAppWindow()->PlatformHandleContextMenu(params);
```
Comment (pozdnyakov) [inline runtime/browser/runtime.cc]: let ui delegate handle it, ui_delegate_->HandleContextMenu(..);

File: runtime/browser/devtools/xwalk_devtools_frontend.h (hunk 7 above)
Comment (darktears) [inline runtime/browser/devtools/xwalk_devtools_frontend.h]: base class is pure virtual so nope.

File: runtime/browser/devtools/xwalk_devtools_frontend.h (hunk 6 above)
Comment (darktears) [inline runtime/browser/devtools/xwalk_devtools_frontend.h]: runtime_window_ doesn't have anything to do with XWalkDevToolsFrontend itself, it just the class is holding a reference to the actual runtime windows we're inspecting.

File: runtime/browser/devtools/xwalk_devtools_manager_delegate.cc (hunk 9 above)
Comment (darktears) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: Not sure, then I would need to pass the value of the port around.

Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: that would be nicer IMO, TCPServerSocketFactory should know about address and shouldn't know about cmd line

File: runtime/browser/runtime_ui_delegate.cc [hunk 13]
```diff
@@ -129,4 +129,12 @@ blink::WebDisplayMode DefaultRuntimeUIDelegate::GetDisplayMode() const {
   return window_params_.display_mode;
 }
 
+bool DefaultRuntimeUIDelegate::HandleContextMenu(
+    const content::ContextMenuParams& params) {
+  if (window_)
```
Comment (pozdnyakov) [inline runtime/browser/runtime_ui_delegate.cc]: return window_ ? window_->PlatformHandleContextMenu(params) : false;

File: runtime/browser/devtools/xwalk_devtools_manager_delegate.cc [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+#include "xwalk/runtime/browser/runtime.h"
+
+using devtools_http_handler::DevToolsHttpHandler;
+
+namespace xwalk {
+
+namespace {
+
+const int kBackLog = 10;
+const char* kLocalHost = "127.0.0.1";
```
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_manager_delegate.cc]: nit: const char kLocalHost[]

File: runtime/browser/devtools/xwalk_devtools_frontend.cc (hunk 5 above)
Comment (pozdnyakov) [inline runtime/browser/devtools/xwalk_devtools_frontend.cc]: I mean when agent_host_ == nullptr

**Comment by pozdnyakov** [comment]
lgtm

**Pull request merged**